### PR TITLE
tests: benchdnn: add capability to import buffers from files

### PR DESCRIPTION
**Pull request by hidefromkgb**
The direct predecessor to this patch was instrumental in resolving [MFDNN-14086](https://jira.devtools.intel.com/browse/MFDNN-14086).

Buffer import capability is a great feature to have when debugging real-world usecases in which the synthetic data generated by `benchdnn` is not always enough to reproduce the error.

### CODE REVIEW

File: tests/benchdnn/utils/fill.cpp [hunk 1]
```diff
@@ -364,3 +364,39 @@ int fill_random_real(dnn_mem_t &mem_ref, const fill_cfg_t &fill_cfg,
     dnn_mem_t dummy;
     return fill_random_real(dummy, mem_ref, nullptr, fill_cfg, dnnl_memory);
 }
+
+int fill_from_file(int exec_arg, dnn_mem_t &mem, dnn_mem_t &mem_ref) {
+    static const char format[] = "File '%s' %s; buffer not imported.\n";
+    auto prefix = benchdnn_getenv_string("BENCHDNN_BUFFER_PREFIX", false);
+    if (prefix.empty()) return FAIL;
+
+    prefix += "." + std::to_string(exec_arg);
```
Comment (echeresh) [inline tests/benchdnn/utils/fill.cpp]: Can we add `.bin` suffix here by default?

I have a branch dumping memory arguments:
https://github.com/uxlfoundation/oneDNN/tree/echeresh/memdump

This would help to make this feature self-contained in oneDNN/benchdnn: oneDNN dumps buffers and then benchdnn loads the buffers.

When adding memory-dumping as an official feature, it would be nice to align with `ONEDNN_JIT_DUMP` including `.bin` suffix: https://uxlfoundation.github.io/oneDNN/dev_guide_inspecting_jit.html

File: tests/benchdnn/utils/fill.cpp [hunk 2]
```diff
@@ -364,3 +364,39 @@ int fill_random_real(dnn_mem_t &mem_ref, const fill_cfg_t &fill_cfg,
...
+    auto file = fopen(prefix.c_str(), "rb");
+    if (!file) {
```
Comment (echeresh) [inline tests/benchdnn/utils/fill.cpp:383]: I suggest to add a quick check on entry to  exit if argument is empty (size is zero).

For scratchpad and workspace we may not have files/data but `fill_from_file()` is still called and will print "cannot be opened" error.

File: tests/benchdnn/bnorm/bnorm.cpp [hunk 3]
```diff
@@ -282,25 +286,33 @@ int prepare_fwd(const prb_t *prb, dnn_mem_map_t &mem_map,
...
 
     return OK;
 }
 
-int prepare_bwd(
-        const prb_t *prb, dnn_mem_map_t &mem_map, dnn_mem_map_t &ref_mem_map) {
+int prepare_bwd(const prb_t *prb, dnn_mem_map_t &mem_map,
+        dnn_mem_map_t &ref_mem_map,
+        const std::unordered_map<int, int> &file_fill_map) {
+    if (file_fill_map.at(DNNL_ARG_DIFF_DST) == OK) return OK;
```
Comment (echeresh) [inline tests/benchdnn/bnorm/bnorm.cpp]: Backward bnorm cases (and maybe RNN?) require forward to be called first. How will it work with this feature?

- Does the user have to provide both sets of buffers?
- If yes, is argument overlapping possible between forward/backward?

If this is not trivial to support we can exclude batchnorm, RNN (and something else maybe?) from the initial support.

**Comment by vpirogov** [comment]
Could you please elaborate on the usage model of this feature? As benchdnn correctness checks rely on certain properties of generated data for stability. How do we plan to test correctness with arbitrary buffers?

File: tests/benchdnn/utils/fill.cpp (hunk 1 above)
Comment (hidefromkgb) [reply tests/benchdnn/utils/fill.cpp]: You mean e.g. `/path/to/prefix.bin.41`?
Will do. I'm okay with that — and it'll also help adding plain-text import in the future (e.g. using a .txt.IDX extension).

File: tests/benchdnn/utils/fill.cpp (hunk 2 above)
Comment (hidefromkgb) [reply tests/benchdnn/utils/fill.cpp:383]: These errors are not fatal, the execution will continue even if some files are absent.
This is by design, just to let the developer know all of the buffers that participate.

File: tests/benchdnn/bnorm/bnorm.cpp (hunk 3 above)
Comment (hidefromkgb) [reply tests/benchdnn/bnorm/bnorm.cpp]: Yup, and pooling as well.
This is rather easy to support by appending an additional suffix for direction, e.g. `/path/to/prefix.fwd_i.bin.41`

**Comment by hidefromkgb** [comment]
> Could you please elaborate on the usage model of this feature? As benchdnn correctness checks rely on certain properties of generated data for stability. How do we plan to test correctness with arbitrary buffers?

@vpirogov This is not for general CI, but rather to give developers a means to debug correctness issues filed by customers, and these are always different case by case, so some correctness issues (due e.g. to a different summation order in CPU vs. GPU, DPAS vs. VNNI, etc.) are inevitable — but there's a difference between 0.41 on CPU vs. 0.42 on GPU and e.g. 0.41 vs. –NaN.

File: tests/benchdnn/utils/fill.cpp (hunk 1 above)
Comment (echeresh) [reply tests/benchdnn/utils/fill.cpp]: I meant at the end: `/path/to/prefix.41.bin`

For JIT dump we put `.bin` prefix at the end - this is an extension to mark a binary file.

File: tests/benchdnn/utils/fill.cpp (hunk 2 above)
Comment (echeresh) [reply tests/benchdnn/utils/fill.cpp:383]: @hidefromkgb "cannot be opened" doesn't sound like a warning or information message. As a user I would likely consider this as "something is missing, I have to do something with it". If you want to keep it - I would suggest to rephrase it to communicate that it's OK for the buffer to be missing.

Comment (hidefromkgb) [reply tests/benchdnn/utils/fill.cpp:383]: @echeresh I've got no issue with "is missing" instead of "cannot be opened". Will change.

**Comment by vpirogov** [comment]
> This is not for general CI, but rather to give developers a means to debug correctness issues filed by customers, and these are always different case by case, so some correctness issues (due e.g. to a different summation order in CPU vs. GPU, DPAS vs. VNNI, etc.) are inevitable — but there's a difference between 0.41 on CPU vs. 0.42 on GPU and e.g. 0.41 vs. –NaN.

I get this part. What I'm saying is that benchdnn correctness check is not equipped to handle arbitrary data and you will likely be getting false positive "test fail" status even for correct results.

File: tests/benchdnn/bnorm/bnorm.cpp (hunk 3 above)
Comment (hidefromkgb) [reply tests/benchdnn/bnorm/bnorm.cpp]: @echeresh BTW aren't the buffer IDs different between different directions? I looked closer and I don't see clashes between buffer IDs across different directions.

**Comment by hidefromkgb** [comment]
> I get this part. What I'm saying is that benchdnn correctness check is not equipped to handle arbitrary data and you will likely be getting false positive "test fail" status even for correct results.

Well, this is what `-v99` and developers' common sense are for. I don't think we need to fundamentally change how benchdnn works just to accommodate for this feature. All of us know that things like machine epsilon exist.

File: tests/benchdnn/utils/fill.cpp (hunk 1 above)
Comment (hidefromkgb) [reply tests/benchdnn/utils/fill.cpp]: OK. LGTM

File: tests/benchdnn/bnorm/bnorm.cpp (hunk 3 above)
Comment (echeresh) [reply tests/benchdnn/bnorm/bnorm.cpp]: @hidefromkgb That makes sense. So you expect the user to provide buffers from forward - can you please update documentation as well to specify that some cases (such as batchnorm) require forward buffers.

Also I think this should be a requirement, if forward buffers are missing, we can't run correctness checks as the buffers won't be coherent. So if missing forward buffers for such cases - should report an error.

Comment (echeresh) [reply tests/benchdnn/bnorm/bnorm.cpp]: > @echeresh BTW aren't the buffer IDs different between different directions? I looked closer and I don't see clashes between buffer IDs across different directions.

What about scratchpads? Also it would be nice to consider future `ONEDNN_MEMORY_DUMP` support - we have to have unique prefixes between cases for that feature so it would be natural to take that into account for this benchdnn feature.

Comment (hidefromkgb) [reply tests/benchdnn/bnorm/bnorm.cpp]: @echeresh Is there a usecase for scratchpad buffer initialization?
Don't get me wrong, directionality is trivial to support, but is it really necessary?

File: tests/benchdnn/utils/fill.cpp (hunk 1 above)
Comment (hidefromkgb) [reply tests/benchdnn/utils/fill.cpp]: Done.

File: tests/benchdnn/utils/fill.cpp (hunk 2 above)
Comment (hidefromkgb) [reply tests/benchdnn/utils/fill.cpp:383]: Done.

**Comment by vpirogov** [comment]
> Well, this is what `-v99` and developers' common sense are for. I don't think we need to fundamentally change how benchdnn works just to accommodate for this feature. All of us know that things like machine epsilon exist.

Thanks, this explain the usage model. As you are introducing a feature available to all benchdnn users we need to add some warnings for users who are not as well versed in machine arithmetic as oneDNN developers.

**Comment by hidefromkgb** [comment]
> Thanks, this explain the usage model. As you are introducing a feature available to all benchdnn users we need to add some warnings for users who are not as well versed in machine arithmetic as oneDNN developers.

@vpirogov Should we maybe hide this under `DNNL_DEV_MODE`? This would alleviate all such concerns.
Anyone who builds oneDNN with this compilation flag basically consents to knowing exactly what they're doing.

**Comment by vpirogov** [comment]
> Should we maybe hide this under `DNNL_DEV_MODE`? This would alleviate all such concerns. Anyone who builds oneDNN with this compilation flag basically consents to know exactly what they're doing.

We have a few more options here:
* Add a warning indicating that false positives are likely
* Stop reporting pas/fail status and show statistics on data difference from reference
I'll defer decision on which approach to take to @dzarukin.

File: tests/benchdnn/bnorm/bnorm.cpp (hunk 3 above)
Comment (echeresh) [reply tests/benchdnn/bnorm/bnorm.cpp]: Yeah, are right about scratchpads. What about mean/variance between forward/backward batch normalization?

https://uxlfoundation.github.io/oneDNN/dev_guide_batch_normalization.html

(I guess here practically we can assume they are the same through between forward/backward)

Anyway, I suggest to take future support for memory dumping in oneDNN into account, say we can have the following files when dumping batch normalization:

```bash
# backward
dnnl_dump_mem.1.gpu.batch_normalization.17.bin
dnnl_dump_mem.1.gpu.batch_normalization.1.bin
dnnl_dump_mem.1.gpu.batch_normalization.49.bin
dnnl_dump_mem.1.gpu.batch_normalization.50.bin
dnnl_dump_mem.1.gpu.batch_normalization.64.bin

# forward
dnnl_dump_mem.3.gpu.batch_normalization.129.bin
dnnl_dump_mem.3.gpu.batch_normalization.145.bin
dnnl_dump_mem.3.gpu.batch_normalization.1.bin
dnnl_dump_mem.3.gpu.batch_normalization.49.bin
dnnl_dump_mem.3.gpu.batch_normalization.50.bin
dnnl_dump_mem.3.gpu.batch_normalization.64.bin
```

Then it could be as simple as specifying two separate prefixes for forward/backward.

Comment (hidefromkgb) [reply tests/benchdnn/bnorm/bnorm.cpp]: @echeresh After a careful consideration I realized that we simply don't need to run both forward and backwards primitives at once if we have the necessary buffers ready — or fail if not given something that otherwise necessitates a forward run.

Comment (echeresh) [reply tests/benchdnn/bnorm/bnorm.cpp]: @hidefromkgb Thanks for checking, if so - that's better. Since this is not the default scenario - can you please test it following the expected usage flow to ensure everything works as expected? If something pops up on the other hand, it's fine to have limitations, if needed.

File: tests/benchdnn/binary/binary.cpp [hunk 4]
```diff
@@ -35,10 +35,11 @@
 
 namespace binary {
 
-int fill_mem(
-        const prb_t *prb, int input_idx, dnn_mem_t &mem_dt, dnn_mem_t &mem_fp) {
+int fill_mem(int input_idx, const prb_t *prb, int exec_arg, dnn_mem_t &mem_dt,
```
Comment (dzarukin) [inline tests/benchdnn/binary/binary.cpp]: Please replace `input_idx` completely with `exec_arg` here.

File: tests/benchdnn/concat/concat.cpp [hunk 5]
```diff
@@ -68,6 +68,7 @@ int fill_src(int input_idx, dnnl_data_type_t dt, dnn_mem_t &mem_dt,
         dnn_mem_t &mem_fp) {
     const auto nelems = mem_fp.nelems();
     if (nelems == 0) return OK;
+    if (fill_from_file(input_idx, mem_dt, mem_fp)) return OK;
```
Comment (dzarukin) [inline tests/benchdnn/concat/concat.cpp]: I'd rename `input_idx` with `exec_arg` here as well.

**Comment by dzarukin** [comment]
make test

File: tests/benchdnn/sum/sum.cpp [hunk 6]
```diff
@@ -68,6 +68,7 @@ dnnl_status_t init_pd(init_pd_args_t<prb_t> &init_pd_args) {
 int fill_src(int input_idx, dnn_mem_t &mem_dt, dnn_mem_t &mem_fp) {
     const auto nelems = mem_dt.nelems();
     if (nelems == 0) return OK;
+    if (fill_from_file(input_idx, mem_dt, mem_fp)) return OK;
```
Comment (dzarukin) [inline tests/benchdnn/sum/sum.cpp]: Ditto.

File: tests/benchdnn/utils/fill.cpp [hunk 7]
```diff
@@ -364,3 +364,43 @@ int fill_random_real(dnn_mem_t &mem_ref, const fill_cfg_t &fill_cfg,
...
+        BENCHDNN_PRINT(2, format, prefix.c_str(), "is missing");
+        return false;
+    }
+    fseek(file, 0, SEEK_END);
+    size_t total = 0, size = ftell(file);
+    if (mem.size() != size) {
+        BENCHDNN_PRINT(2, format, prefix.c_str(),
+                "differs in size from the buffer's memory descriptor");
+        return false;
```
Comment (dzarukin) [inline tests/benchdnn/utils/fill.cpp]: I suggest to inform the user at 0 level about the ill-formed file and abort the execution.
```suggestion
        BENCHDNN_PRINT(0, format, prefix.c_str(),
                "differs in size from the buffer's memory descriptor");
        fclose(file);
        SAFE_V(FAIL);
```

File: tests/benchdnn/utils/fill.cpp [hunk 8]
```diff
@@ -364,3 +364,43 @@ int fill_random_real(dnn_mem_t &mem_ref, const fill_cfg_t &fill_cfg,
...
+
+bool fill_from_file(int exec_arg, dnn_mem_t &mem, dnn_mem_t &ref_mem) {
+    static const char format[] = "File '%s' %s; buffer not imported.\n";
+    auto prefix = buffer_prefix;
+    if (prefix.empty()) return false;
+
+    prefix += "." + std::to_string(exec_arg) + ".bin";
+    auto file = fopen(prefix.c_str(), "rb");
+    if (!file) {
+        BENCHDNN_PRINT(2, format, prefix.c_str(), "is missing");
```
Comment (dzarukin) [inline tests/benchdnn/utils/fill.cpp]:
```suggestion
        BENCHDNN_PRINT(2, format, prefix.c_str(), "is not found");
```

File: tests/benchdnn/utils/fill.cpp [hunk 9]
```diff
@@ -364,3 +364,43 @@ int fill_random_real(dnn_mem_t &mem_ref, const fill_cfg_t &fill_cfg,
...
+        BENCHDNN_PRINT(2, format, prefix.c_str(),
+                "differs in size from the buffer's memory descriptor");
+        return false;
+    }
+    fseek(file, 0, SEEK_SET);
+    for (size_t read = ~0; read && (total < size); total += read)
+        read = fread(
+                static_cast<uint8_t *>(mem) + total, 1, size - total, file);
+    if (total != size) {
+        BENCHDNN_PRINT(2, format, prefix.c_str(), "cannot be read correctly");
```
Comment (dzarukin) [inline tests/benchdnn/utils/fill.cpp]: Ditto as for ill-formed file. Read incorrectly - must provide guidance what could go wrong and abort.

File: tests/benchdnn/utils/fill.cpp [hunk 10]
```diff
@@ -364,3 +364,43 @@ int fill_random_real(dnn_mem_t &mem_ref, const fill_cfg_t &fill_cfg,
...
+        return false;
+    }
+    fclose(file);
+    if (has_bench_mode_bit(mode_bit_t::corr)) {
```
Comment (dzarukin) [inline tests/benchdnn/utils/fill.cpp]: Is it possible to change this to simply `if (ref_mem) {`?

File: tests/benchdnn/utils/fill.cpp [hunk 11]
```diff
@@ -364,3 +364,43 @@ int fill_random_real(dnn_mem_t &mem_ref, const fill_cfg_t &fill_cfg,
...
+        if (ref_mem.reorder(mem) != OK) {
+            BENCHDNN_PRINT(2, format, prefix.c_str(), "cannot be reordered");
```
Comment (dzarukin) [inline tests/benchdnn/utils/fill.cpp]: I'd abort this as well as two previous cases.

File: tests/benchdnn/binary/binary.cpp (hunk 4 above)
Comment (hidefromkgb) [reply tests/benchdnn/binary/binary.cpp]: Done.

File: tests/benchdnn/concat/concat.cpp (hunk 5 above)
Comment (hidefromkgb) [reply tests/benchdnn/concat/concat.cpp]: Done.

File: tests/benchdnn/sum/sum.cpp (hunk 6 above)
Comment (hidefromkgb) [reply tests/benchdnn/sum/sum.cpp]: Done.

File: tests/benchdnn/utils/fill.cpp (hunk 8 above)
Comment (hidefromkgb) [reply tests/benchdnn/utils/fill.cpp]: Done.

File: tests/benchdnn/utils/fill.cpp (hunk 7 above)
Comment (hidefromkgb) [reply tests/benchdnn/utils/fill.cpp]: Done.

File: tests/benchdnn/utils/fill.cpp (hunk 9 above)
Comment (hidefromkgb) [reply tests/benchdnn/utils/fill.cpp]: Done.

File: tests/benchdnn/utils/fill.cpp (hunk 10 above)
Comment (hidefromkgb) [reply tests/benchdnn/utils/fill.cpp]: Done.

File: tests/benchdnn/utils/fill.cpp (hunk 11 above)
Comment (hidefromkgb) [reply tests/benchdnn/utils/fill.cpp]: Done.

**Comment by hidefromkgb** [comment]
make test
